src/ucp/tag/offload.c

```diff
@@ -117,14 +118,19 @@ UCS_PROFILE_FUNC_VOID(ucp_tag_offload_completed,
                                     UCP_RECV_DESC_FLAG_EAGER_OFFLOAD);
     }
 
-    if (req->recv.tag.rdesc != NULL) {
+    if (ucs_unlikely(inline_data != NULL)) {
+        status = ucp_request_recv_data_unpack(req, inline_data, length, 0, 1);
+    } else if (req->recv.tag.rdesc != NULL) {
         status = ucp_request_recv_data_unpack(req, req->recv.tag.rdesc + 1,
                                               length, 0, 1);
-        ucs_mpool_put_inline(req->recv.tag.rdesc);
     } else {
         ucp_request_recv_buffer_dereg(req);
     }
 
+    if (req->recv.tag.rdesc != NULL) {
+        ucs_mpool_put_inline(req->recv.tag.rdesc);
```

Comment:
why not invoke it at line 125?

Reply:
we need to release it even when inline_data != NULL right?

Reply:
yes, you are right

Reply:
maybe better do it both in line 125 and also after 122, to avoid extra branch